app/src/main/java/org/mozilla/fenix/home/blocklist/BlocklistMiddleware.kt

```diff
@@ -0,0 +1,140 @@
+                }
+                is HomeFragmentAction.RemoveRecentTab -> {
+                    if (action.recentTab is RecentTab.Tab) {
+                        val url = action.recentTab.state.content.url
+                        val updatedBlocklist = blocklist + url
+                        settings.homescreenBlocklist = updatedBlocklist
+                        state.filteredByBlocklist(updatedBlocklist).toChangeAction()
+                    } else {
+                        action
+                    }
+                }
+                is HomeFragmentAction.RemoveRecentBookmark -> {
+                    action.recentBookmark.url?.let {
+                        val updatedBlocklist = blocklist + it
+                        settings.homescreenBlocklist = updatedBlocklist
+                        state.filteredByBlocklist(updatedBlocklist).toChangeAction()
+                    } ?: action
+                }
+                is HomeFragmentAction.RemoveRecentHistoryHighlight -> {
+                    val url = action.highlightUrl
+                    val updatedBlocklist = blocklist + url
+                    settings.homescreenBlocklist = updatedBlocklist
+                    state.filteredByBlocklist(updatedBlocklist).toChangeAction()
+                }
+                else -> action
+            }
+        )
+    }
+
+    private fun HomeFragmentState.toChangeAction() = HomeFragmentAction.Change(
+        topSites = topSites,
+        recentTabs = recentTabs,
+        recentBookmarks = recentBookmarks,
+        recentHistory = recentHistory,
+        mode = mode,
+        collections = collections,
+        tip = tip,
+        showCollectionPlaceholder = showCollectionPlaceholder
+    )
+
+    fun filterState(state: HomeFragmentState): HomeFragmentState =
+        state.filteredByBlocklist(blocklist)
+
+    private fun HomeFragmentState.filteredByBlocklist(blocklist: Set<String>) = this.copy(
+        recentBookmarks = recentBookmarks.filteredByBlocklist(blocklist),
+        recentTabs = recentTabs.filteredByBlocklist(blocklist),
+        recentHistory = recentHistory.filteredByBlocklist(blocklist)
+    )
+
+    private fun Set<String>.containsUrl(url: String): Boolean {
+        val strippedUrl = url.stripProtocolAndCommonSubdomains()
+        return this.any { it.stripProtocolAndCommonSubdomains() == strippedUrl }
+    }
+
+    private fun String.stripProtocolAndCommonSubdomains(): String {
+        val protocolStripped = this.substringAfter("://")
+        // This kind of stripping allows us to match "twitter" to "mobile.twitter.com".
+        // Borrowed from DomainMatcher in A-C
+        val domainsToStrip = listOf("www", "mobile", "m")
+
+        domainsToStrip.forEach { domain ->
+            if (protocolStripped.startsWith("$domain.")) {
+                return protocolStripped.substring(domain.length + 1)
+            }
+        }
+        return protocolStripped
+    }
+
+    @JvmName("filterRecentBookmark")
```

Comment:
Do we need theses `@JvmName`? I am mostly curious because we don't have a history of its usage and I don't know much about their usage.

Reply:
I assume this is a JVM type erasure thing, where the boxed type in the various `List<*>` extensions defined here isn't known and the JVM needs a way to differentiate between them. 

Alternatively I could define them as non-extension functions with names like `filterRecentBookmarks` and drop the `blocklist` naming convention.

Reply:
Oh I see. Seems okay.